packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/Drawing/helpers/withCoordsFromStream.js

```diff
@@ -0,0 +1,115 @@
+import { inject } from 'mobx-react'
+import PropTypes from 'prop-types'
+import React, { Component } from 'react'
+
+function storeMapper (stores) {
+  const {
+    eventStream
+  } = stores.classifierStore.drawing
+  return {
+    eventStream
+  }
+}
+
+function withCoordsFromStream (WrappedComponent) {
+  @inject(storeMapper)
+  class CoordsFromStream extends Component {
+    constructor () {
+      super()
+
+      this.state = {
+        activeCoordinates: []
+      }
+
+      this.firstSubscription = null
+      this.secondSubscription = null
+
+      this.finishDrawing = this.finishDrawing.bind(this)
+      this.startDrawing = this.startDrawing.bind(this)
+    }
+
+    componentDidMount () {
+      this.startDrawing()
+    }
+
+    componentWillUnmount () {
+      if (this.firstSubscription) {
+        this.firstSubscription.unsubscribe()
+      }
+      if (this.secondSubscription) {
+        this.secondSubscription.unsubscribe()
+      }
+    }
+
+    startDrawing () {
+      if (this.props.active) {
+        const { eventStream } = this.props
+        this.firstSubscription = eventStream.subscribe(event => {
+          if (event.type === 'pointerdown') {
+            this.secondSubscription = eventStream.subscribe(event => {
+              const svgCoordinateEvent = this.convertEvent(event)
+              const activeCoordinates = Array.from(this.state.activeCoordinates)
+              activeCoordinates.push(svgCoordinateEvent)
+              this.setState({ activeCoordinates })
+            })
+          }
+        })
+      }
+    }
+
+    convertEvent (event) {
+      const type = event.type
+
+      const clientX = event.clientX
+      const clientY = event.clientY
+      const svgEventOffset = this.getEventOffset(clientX, clientY)
+
+      const svgCoordinateEvent = {
+        type,
+        x: svgEventOffset.x,
+        y: svgEventOffset.y
+      }
+
+      return svgCoordinateEvent
+    }
+
+    getEventOffset (x, y) {
+      const { svg } = this.props
+
+      const svgEvent = svg.createSVGPoint()
+      svgEvent.x = x
+      svgEvent.y = y
+      const svgEventOffset = svgEvent.matrixTransform(svg.getScreenCTM().inverse())
+
+      return svgEventOffset
+    }
+
+    finishDrawing (coordinates) {
+      this.secondSubscription.unsubscribe()
+      this.props.storeMark(coordinates)
+    }
+
+    render () {
+      const { ...props } = this.props
+      const { activeCoordinates } = this.state
+
+      return (
+        <WrappedComponent
+          coordinates={activeCoordinates}
```

Comment:
`activeCoordinates` is an array of simplified event objects that consist of properties `x`, `y`, and the event type (i.e. `pointerdown`, `pointermove`, `pointerup`). 

The idea is that each drawing tool has unique functionality and knows when to conclude drawing, so let's send the minimum amount of event data and a callback to finish drawing to the drawing tool component, then the drawing tool component will do what it needs to do with the simplified events and let parent components/containers know when that particular drawing tool mark has been completed.

Reply:
Does this mean that the drawing tool itself doesn't receive a stream, but a reference to an array held in state as a prop?

If so, it might be an idea to simply give the drawing tool its simplified stream (again, `switchMap` is likely your friend), as `setState` calls are batched, I can imagine lots of re-renders happening, and the drawing tool code itself could be simplified by using a stream.